### PR TITLE
use nonuser thread access tokens to grant keep view permissions

### PR DESCRIPTION
**Pull request by camhashemi**
@ryanpbrewster @leogrim @andrewconner 

this is kind of a pain for such a low-usage feature. @ryanpbrewster mentioned creating a nonuser -> keep table for stuff like this, we agreed it's not needed for now.

### CODE REVIEW

**Comment by camhashemi** [comment]
@ryanpbrewster if the user signs up with this access code, we'll convert the non-user thread, add the user to the keep.

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/DeepLinkRouter.scala [hunk 1]
```diff
@@ -58,11 +58,15 @@ class DeepLinkRouterImpl @Inject() (
...
-      }
+    val keepPageOpt = for {
+      _ <- Some(()) if redirectToKeepPage
+      keepPubId <- (data \ DeepLinkField.KeepId).asOpt[PublicId[Keep]]
+      keepId <- Keep.decodePublicId(keepPubId).toOption
+      keep <- db.readOnlyReplica(implicit s => keepRepo.getOption(keepId))
+    } yield {
+      val accessTokenOpt = (data \ DeepLinkField.AuthToken).asOpt[String]
+      keep.path.relative + accessTokenOpt.map(token => s"?authToken=$token").getOrElse("")
+    }
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/DeepLinkRouter.scala]: :boom:, nice

**Comment by ryanpbrewster** [comment]
I mean, you don't need me to tell you how unpleasant this must be to do. Nice work, though.

Looks fine to me. Long-term I'd like to have a `KeepToNonUser` model that we can just check directly (thus allowing us to drop all the Eliza queries). However, I'm not sure we're really going to be able to do away with the `Tag4` stuff, ever.

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/DeepLinkRouter.scala (hunk 1 above)
Comment (camhashemi) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/DeepLinkRouter.scala]: style jacker alert

File: kifi-backend/modules/common/app/com/keepit/common/routes/Router.scala [hunk 2]
```diff
@@ -267,6 +267,7 @@ object Eliza extends Service {
     def getMessagesOnKeep = ServiceRoute(POST, "/internal/eliza/getMessagesOnKeep")
     def editMessage() = ServiceRoute(POST, "/internal/eliza/editMessage")
     def deleteMessage() = ServiceRoute(POST, "/internal/eliza/deleteMessage")
+    def keepHasAccessToken(keepId: Id[Keep], accessToken: String) = ServiceRoute(GET, "/internal/eliza/keepHasAccessToken")
```
Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/common/routes/Router.scala:270]: Doesn't this need a `Param` or two?

File: kifi-backend/modules/eliza/app/com/keepit/eliza/commanders/ElizaDiscussionCommander.scala [hunk 3]
```diff
@@ -206,6 +209,11 @@ class ElizaDiscussionCommanderImpl @Inject() (
   def deleteMessage(messageId: Id[ElizaMessage]): Unit = db.readWrite { implicit s =>
     messageRepo.deactivate(messageRepo.get(messageId))
   }
+
+  def keepHasAccessToken(keepId: Id[Keep], accessToken: ThreadAccessToken): Boolean = db.readOnlyMaster { implicit s =>
+    nonUserThreadRepo.getByAccessToken(accessToken).exists(_.keepId == keepId)
```
Comment (ryanpbrewster) [inline kifi-backend/modules/eliza/app/com/keepit/eliza/commanders/ElizaDiscussionCommander.scala:214]: Clever. I would have done this in a way dumber way. :+1:

**Pull request merged**